modules/auth/sso/sso.go

```diff
@@ -0,0 +1,181 @@
+// for users that have already signed in.
+var ssoMethods []SingleSignOn = []SingleSignOn{
+	&OAuth2{},
+	&Session{},
+	&ReverseProxy{},
+	&Basic{},
+}
+
+// The purpose of the following three function variables is to let the linter know that
+// those functions are not dead code and are actually being used
+var (
+	_ = isPublicResource
+	_ = isPublicPage
+	_ = handleSignIn
+)
+
+// Methods returns the instances of all registered SSO methods
+func Methods() []SingleSignOn {
+	return ssoMethods
+}
+
+// Register adds the specified instance to the list of available SSO methods
+func Register(method SingleSignOn) {
+	ssoMethods = append(ssoMethods, method)
+}
+
+// Init should be called exactly once when the application starts to allow SSO plugins
+// to allocate necessary resources
+func Init() {
+	for _, method := range Methods() {
+		err := method.Init()
+		if err != nil {
+			log.Error("Could not initialize '%s' SSO method, error: %s", reflect.TypeOf(method).String(), err)
+		}
+	}
+}
+
+// Free should be called exactly once when the application is terminating to allow SSO plugins
+// to release necessary resources
+func Free() {
+	for _, method := range Methods() {
+		err := method.Free()
+		if err != nil {
+			log.Error("Could not free '%s' SSO method, error: %s", reflect.TypeOf(method).String(), err)
+		}
+	}
+}
+
+// SessionUser returns the user object corresponding to the "uid" session variable.
+func SessionUser(sess session.Store) *models.User {
+	// Get user ID
+	uid := sess.Get("uid")
+	if uid == nil {
+		return nil
+	}
+	id, ok := uid.(int64)
+	if !ok {
+		return nil
+	}
+
+	// Get user object
+	user, err := models.GetUserByID(id)
+	if err != nil {
+		if !models.IsErrUserNotExist(err) {
+			log.Error("GetUserById: %v", err)
+		}
+		return nil
+	}
+	return user
+}
+
+// isAPIPath returns true if the specified URL is an API path
+func isAPIPath(ctx *macaron.Context) bool {
+	return strings.HasPrefix(ctx.Req.URL.Path, "/api/")
+}
+
+// isAttachmentDownload check if request is a file download (GET) with URL to an attachment
+func isAttachmentDownload(ctx *macaron.Context) bool {
+	return strings.HasPrefix(ctx.Req.URL.Path, "/attachments/") && ctx.Req.Method == "GET"
+}
+
+// isPublicResource checks if the url is of a public resource file that should be served
+// without authentication (eg. the Web App Manifest, the Service Worker script or the favicon)
+func isPublicResource(ctx *macaron.Context) bool {
+	path := strings.TrimSuffix(ctx.Req.URL.Path, "/")
+	return path == "/robots.txt" ||
+		path == "/favicon.ico" ||
+		path == "/favicon.png" ||
+		path == "/manifest.json" ||
+		path == "/serviceworker.js"
+}
+
+// isPublicPage checks if the url is of a public page that should not require authentication
+func isPublicPage(ctx *macaron.Context) bool {
+	path := strings.TrimSuffix(ctx.Req.URL.Path, "/")
+	homePage := strings.TrimSuffix(setting.AppSubURL, "/")
+	currentURL := homePage + path
+	return currentURL == homePage ||
```

Comment:
The SSPI authentication is executed inside the [Contexter](https://github.com/go-gitea/gitea/blob/9f67df8cd75e7bceb3ee8077ccc4c84b455dfdf6/modules/context/context.go#L303) macaron middleware, which happens before `reqSignIn` determines if the user has to sign in (that happens just before the page handler is executed, inside the last middleware in the queue).

Initially I planned for SSPI to use a flag, but in the opposite way - set a flag if SSPI authentication fails and later check that flag inside the Toggle middleware, returning 403 if the flag was set. I could check the value of that flag just after the `if !ctx.IsSigned {` line):
https://github.com/go-gitea/gitea/blob/171b3598778a1ecd0a921c71ed6755bfef68f7f0/modules/context/auth.go#L76-L78

I gave up on that plan, mainly because I did not want to split the authentication logic at two places, so far apart, with other middlewares in between. Will give that some more tought.

Reply:
Simplified that part by removing `isPublicResource()` and `isPublicPage()` functions. Instead, changed the SSPI auth method to perform/require authentication only on its own login page (`/user/login?auth_with_sspi=1`), API paths and download URLs.

For all other URLs that require authentication (eg. pages with reqSignIn) it relies on the `Toggle` middleware to redirect anonymous users to the login page. `Toggle` performs authorization anyway, so this is actually not a drawback.

Fixed with commit https://github.com/go-gitea/gitea/pull/8463/commits/6b6e40544284c6403f9f3467ba3f148b30b6886c